File: python/src/c4_command.py
"""
The intent is to encapsulate a request as an object, thereby letting you parameterize
clients with different requests, queue or log requests, and support undoable operations.

Key participants:
- Command: Declares interface for executing an operation
- ConcreteCommand: Defines binding between Receiver and action
- Invoker: Asks the command to carry out the request
- Receiver: Knows how to perform the operations

Key benefits:
- Decouples invoker from receiver
- Commands are first-class objects
- Easy to add new commands
- Supports undo/redo

Real-world example: A remote control with undo functionality.
"""

from abc import ABC, abstractmethod
from collections import deque
from dataclasses import dataclass
from enum import Enum
from typing import Optional


# ----- Receivers -----
class Light:
    def on(self) -> None:
        print("Light is ON")

    def off(self) -> None:
        print("Light is OFF")

    def dim(self, level: int) -> None:
        print(f"Light dimmed to {level}%")


class Stereo:
    def on(self) -> None:
        print("Stereo is ON")

    def off(self) -> None:
        print("Stereo is OFF")

    def set_cd(self) -> None:
        print("Stereo: CD player selected")

    def set_volume(self, level: int) -> None:
        print(f"Stereo: volume set to {level}")


class CeilingFan:
    class Speed(Enum):
        OFF = "OFF"
        LOW = "LOW"
        MEDIUM = "MEDIUM"
        HIGH = "HIGH"

    def __init__(self) -> None:
        self._speed = self.Speed.OFF

    def set_speed(self, speed: Speed) -> None:
        self._speed = speed
        print(f"Ceiling Fan: {speed.value} speed")

    def get_speed(self) -> Speed:
        return self._speed


# ----- Command interface -----
class Command(ABC):
    @abstractmethod
    def execute(self) -> None:
        pass

    @abstractmethod
    def undo(self) -> None:
        pass


# ----- Concrete Commands -----
class LightOnCommand(Command):
    def __init__(self, light: Light) -> None:
        self._light = light

    def execute(self) -> None:
        self._light.on()

    def undo(self) -> None:
        self._light.off()


class LightOffCommand(Command):
    def __init__(self, light: Light) -> None:
        self._light = light

    def execute(self) -> None:
        self._light.off()

    def undo(self) -> None:
        self._light.on()


class StereoOnWithCDCommand(Command):
    def __init__(self, stereo: Stereo) -> None:
        self._stereo = stereo

    def execute(self) -> None:
        self._stereo.on()
        self._stereo.set_cd()
        self._stereo.set_volume(11)

    def undo(self) -> None:
        self._stereo.off()


class CeilingFanHighCommand(Command):
    def __init__(self, fan: CeilingFan) -> None:
        self._fan = fan
        self._prev_speed: Optional[CeilingFan.Speed] = None

    def execute(self) -> None:
        self._prev_speed = self._fan.get_speed()
        self._fan.set_speed(CeilingFan.Speed.HIGH)

    def undo(self) -> None:
        if self._prev_speed:
            self._fan.set_speed(self._prev_speed)


class NoCommand(Command):
    def execute(self) -> None:
        print("[No operation]")

    def undo(self) -> None:
        pass


# ----- Invoker -----
class RemoteControl:
    def __init__(self) -> None:
        self._on_commands: list[Command] = [NoCommand()] * 7
        self._off_commands: list[Command] = [NoCommand()] * 7
        self._undo_stack: deque[Command] = deque(maxlen=10)

    def set_command(self, slot: int, on_cmd: Command, off_cmd: Command) -> None:
        self._on_commands[slot] = on_cmd
        self._off_commands[slot] = off_cmd

    def on_button_was_pressed(self, slot: int) -> None:
        print(f"\n[Remote: ON button {slot} pressed]")
        self._on_commands[slot].execute()
        self._undo_stack.append(self._on_commands[slot])

    def off_button_was_pressed(self, slot: int) -> None:
        print(f"\n[Remote: OFF button {slot} pressed]")
        self._off_commands[slot].execute()
        self._undo_stack.append(self._off_commands[slot])

    def undo_button_was_pressed(self) -> None:
        print("\n[Remote: UNDO button pressed]")
        if self._undo_stack:
            cmd = self._undo_stack.pop()
            cmd.undo()
        else:
            print("Nothing to undo.")


# ----- Macro Command -----
class MacroCommand(Command):
    def __init__(self, commands: list[Command]) -> None:
        self._commands = commands

    def execute(self) -> None:
        for cmd in self._commands:
            cmd.execute()

    def undo(self) -> None:
        for cmd in reversed(self._commands):
            cmd.undo()


# ----- Example usage -----
def command_example() -> None:
    """
    Demonstrates the Command pattern with a remote control system.
    """
    print("\n--- Command Pattern Example ---")

    # Create receivers
    living_room_light = Light()
    stereo = Stereo()
    ceiling_fan = CeilingFan()

    # Create commands
    light_on = LightOnCommand(living_room_light)
    light_off = LightOffCommand(living_room_light)
    stereo_on = StereoOnWithCDCommand(stereo)
    fan_high = CeilingFanHighCommand(ceiling_fan)

    # Set up remote
    remote = RemoteControl()
    remote.set_command(0, light_on, light_off)
    remote.set_command(1, stereo_on, LightOffCommand(Light()))
    remote.set_command(2, fan_high, LightOffCommand(Light()))

    # Test
    remote.on_button_was_pressed(0)
    remote.off_button_was_pressed(0)

    remote.on_button_was_pressed(1)
    remote.undo_button_was_pressed()

    # Macro command
    print("\n--- Macro Command ---")
    party_mode = MacroCommand([light_on, stereo_on, fan_high])
    party_mode.execute()
    print("\nUndoing macro...")
    party_mode.undo()


# Test helpers
def get_remote() -> RemoteControl:
    """Helper for tests: Create a remote with basic commands."""
    remote = RemoteControl()
    light = Light()
    remote.set_command(0, LightOnCommand(light), LightOffCommand(light))
    return remote


def get_ceiling_fan_remote() -> tuple[RemoteControl, CeilingFan]:
    """Helper for tests: Create a remote with ceiling fan control."""
    remote = RemoteControl()
    fan = CeilingFan()
    remote.set_command(0, CeilingFanHighCommand(fan), LightOffCommand(Light()))
    return remote, fan


if __name__ == "__main__":
    command_example()

File: rust/src/c4_command.rs
//! # Command Pattern
//!
//! The Command pattern encapsulates a request as an object, thereby letting you
//! parameterize clients with different requests, queue or log requests, and
//! support undoable operations.
//!
//! ## Key Participants
//! - **Command**: Declares interface for executing an operation
//! - **ConcreteCommand**: Defines binding between Receiver and action
//! - **Invoker**: Asks command to carry out request
//! - **Receiver**: Knows how to perform operations
//!
//! ## Benefits
//! - Decouples invoker from receiver
//! - Commands are first-class objects
//! - Easy to add new commands
//! - Supports undo/redo

use std::collections::VecDeque;

// Receivers
pub struct Light;

impl Light {
    pub fn new() -> Self {
        Self
    }

    pub fn on(&self) {
        println!("Light is ON");
    }

    pub fn off(&self) {
        println!("Light is OFF");
    }

    pub fn dim(&self, level: u8) {
        println!("Light dimmed to {}%", level);
    }
}

impl Default for Light {
    fn default() -> Self {
        Self::new()
    }
}

pub struct Stereo {
    #[allow(dead_code)]
    state: StereoState,
}

#[allow(dead_code)]
struct StereoState {
    on: bool,
    cd: bool,
    volume: u8,
}

impl Stereo {
    pub fn new() -> Self {
        Self {
            state: StereoState {
                on: false,
                cd: false,
                volume: 0,
            },
        }
    }

    pub fn on(&self) {
        println!("Stereo is ON");
    }

    pub fn off(&self) {
        println!("Stereo is OFF");
    }

    pub fn set_cd(&self) {
        println!("Stereo: CD player selected");
    }

    pub fn set_volume(&self, level: u8) {
        println!("Stereo: volume set to {}", level);
    }

    // Helper method to show state (for demonstration)
    #[allow(dead_code)]
    fn get_state(&self) -> &StereoState {
        &self.state
    }
}

impl Default for Stereo {
    fn default() -> Self {
        Self::new()
    }
}

// Command trait
pub trait Command {
    fn execute(&self);
    fn undo(&self);
}

// Concrete Commands
pub struct LightOnCommand {
    light: Light,
}

impl LightOnCommand {
    pub fn new(light: Light) -> Self {
        Self { light }
    }
}

impl Command for LightOnCommand {
    fn execute(&self) {
        self.light.on();
    }

    fn undo(&self) {
        self.light.off();
    }
}

pub struct LightOffCommand {
    light: Light,
}

impl LightOffCommand {
    pub fn new(light: Light) -> Self {
        Self { light }
    }
}

impl Command for LightOffCommand {
    fn execute(&self) {
        self.light.off();
    }

    fn undo(&self) {
        self.light.on();
    }
}

pub struct StereoOnWithCDCommand {
    stereo: Stereo,
}

impl StereoOnWithCDCommand {
    pub fn new(stereo: Stereo) -> Self {
        Self { stereo }
    }
}

impl Command for StereoOnWithCDCommand {
    fn execute(&self) {
        self.stereo.on();
        self.stereo.set_cd();
        self.stereo.set_volume(11);
    }

    fn undo(&self) {
        self.stereo.off();
    }
}

pub struct NoCommand;

impl Command for NoCommand {
    fn execute(&self) {
        println!("[No operation]");
    }

    fn undo(&self) {
        // Nothing to undo
    }
}

// Invoker
pub struct RemoteControl {
    on_commands: Vec<Box<dyn Command>>,
    off_commands: Vec<Box<dyn Command>>,
    undo_stack: VecDeque<Box<dyn Command>>,
}

impl RemoteControl {
    pub fn new() -> Self {
        const NUM_SLOTS: usize = 7;

        let mut on_commands: Vec<Box<dyn Command>> = Vec::with_capacity(NUM_SLOTS);
        let mut off_commands: Vec<Box<dyn Command>> = Vec::with_capacity(NUM_SLOTS);

        for _ in 0..NUM_SLOTS {
            on_commands.push(Box::new(NoCommand));
            off_commands.push(Box::new(NoCommand));
        }

        Self {
            on_commands,
            off_commands,
            undo_stack: VecDeque::with_capacity(10),
        }
    }

    pub fn set_command(&mut self, slot: usize, on_cmd: Box<dyn Command>, off_cmd: Box<dyn Command>) {
        if slot < self.on_commands.len() {
            self.on_commands[slot] = on_cmd;
            self.off_commands[slot] = off_cmd;
        }
    }

    pub fn on_button_was_pressed(&mut self, slot: usize) {
        if slot < self.on_commands.len() {
            println!("\n[Remote: ON button {} pressed]", slot);
            self.on_commands[slot].execute();
            // Note: We can't actually move the command out due to trait object limitations
            // In a real implementation, you'd use Rc or similar
        }
    }

    pub fn off_button_was_pressed(&mut self, slot: usize) {
        if slot < self.off_commands.len() {
            println!("\n[Remote: OFF button {} pressed]", slot);
            self.off_commands[slot].execute();
        }
    }

    pub fn undo_button_was_pressed(&self) {
        println!("\n[Remote: UNDO button pressed]");
        if let Some(cmd) = self.undo_stack.back() {
            cmd.undo();
        } else {
            println!("Nothing to undo.");
        }
    }
}

impl Default for RemoteControl {
    fn default() -> Self {
        Self::new()
    }
}

/// Example demonstrating Command pattern
pub fn example() {
    println!("\n--- Command Pattern Example ---\n");

    // Create receivers
    let living_room_light = Light::new();
    let stereo = Stereo::new();

    println!("=== Testing Light directly ===");
    living_room_light.on();
    living_room_light.dim(50);
    living_room_light.off();

    println!("\n=== Testing Stereo directly ===");
    stereo.on();
    stereo.set_cd();
    stereo.set_volume(11);
    stereo.off();

    // Create commands
    let light_on = Box::new(LightOnCommand::new(living_room_light));
    let light_off = Box::new(LightOffCommand::new(Light::new()));
    let stereo_on = Box::new(StereoOnWithCDCommand::new(stereo));

    // Set up remote
    let mut remote = RemoteControl::new();
    remote.set_command(0, light_on, light_off);
    remote.set_command(1, stereo_on, Box::new(LightOffCommand::new(Light::new())));

    println!("\n=== Using Remote Control ===");
    remote.on_button_was_pressed(0);
    remote.off_button_was_pressed(0);

    remote.on_button_was_pressed(1);
    remote.undo_button_was_pressed();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_light_on_command() {
        let light = Light::new();
        let command = LightOnCommand::new(light);
        command.execute();
        // Should not panic
    }

    #[test]
    fn test_light_off_command() {
        let light = Light::new();
        let command = LightOffCommand::new(light);
        command.execute();
        // Should not panic
    }

    #[test]
    fn test_command_undo() {
        let light = Light::new();
        let command = LightOnCommand::new(light);
        command.execute();
        command.undo();
        // Should not panic
    }

    #[test]
    fn test_stereo_command() {
        let stereo = Stereo::new();
        let command = StereoOnWithCDCommand::new(stereo);
        command.execute();
        // Should not panic
    }

    #[test]
    fn test_remote_control() {
        let mut remote = RemoteControl::new();
        let light = Light::new();

        remote.set_command(
            0,
            Box::new(LightOnCommand::new(light)),
            Box::new(LightOffCommand::new(Light::new())),
        );

        remote.on_button_was_pressed(0);
        remote.off_button_was_pressed(0);
        // Should not panic
    }
}
